### ISSUE TITLE
[FEATURE] Add more providers / other orchestrators

### ISSUE THREAD

**Issue by maxlerebourg** [ckotzbauer/sbom-operator#338]
Hey,
I love this project, I already rewrite it for nomad on my personal gitlab so maybe I can help you if possible.

**Is your feature request related to a problem? Please describe.** 🐛
I would like this project to be able to be used on Docker swarm or Nomad to get SBOM from multiple providers.

**Describe the solution you'd like** ✨
In the code, a lot of things are related to Kubernetes (even naming 🥲), maybe create a providers folder like for targets and put all kubernetes related things in the right file (In my rewrite, I totally deleted kubernetes related things 😉 that's why I can't submit a PR).

**Additional context**
To get images from multiple provider, you can read [Diun's code](https://github.com/crazy-max/diun/tree/master/internal/provider) which does almost the same job, get the image + version from providers.
Annotations are not available yet on nomad, so I don't have the exact same provider API than for Kubernetes.

**Comment by ckotzbauer** [comment]
Hey,
thanks for opening this issue and the very positive feedback.
True, right now, the operator is fully based on a kubernetes infrastructure. To be honest, I never had a plan to abstract the kubernetes stuff away to support other runtimes as nomad or swarm or even lambdas. But I wouldn't want to deny or reject your request. I think for some parts of the community, this could be very helpful.

One really important question is how the application can be tested in a good way. Right now, there are unit-tests for the main logic parts, no full integration tests. If I do changes which need a short "live-verification" I test it in my private cluster. But I don't have a swarm or nomad-infrastructure.

It would also be important to have a comparison of possible runtimes with kubernetes to see which features can be used in a different way or can't be used at all. I never worked with nomad before and swarm is already a very long time ago :).

**Comment by maxlerebourg** [comment]
I could write some deployment scripts for testing inside the deployment folder, and possibly, I could take the lead on issue for Nomad, I have a open source project with [exemple](https://github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/tree/main/exemples) folder, I could reproduce it for helping you in a PR.

The comparison could be done for only the features you use:
- run job-image
- get docker image + version
- annotate pod on runtime (coundn't be done on nomad exept by setting variables at node/client level)
I think it's all 

(We could discuss about the second project too, why need kubernetes api inside vulnerability-operator 🥲, but it's for another issue).

**Comment by ckotzbauer** [comment]
> why need kubernetes api inside vulnerability-operator smiling_face_with_tear

It is not that helpful when you only know in which images your CVEs are. You don't want to search for all occurrences of a specifc image in your infrastructure. That's the reason why context information (like Container, Pod or Namespace) are provided.

**Comment by ckotzbauer** [comment]
I just read a bit about nomad. I think it would be possible to port the most important features to nomad. But I would do two assumptions:
- The supported driver would only be "docker", maybe "containerd" or "podman" in the future
- The authentication for private images needs to be done in the task file (`auth` block). Doing the authentication via docker-config or plugin-config won't be possible.

Would this fit your needs?

**Comment by maxlerebourg** [comment]
The first assumtion is what I had in mind, so that's not a problem.
The second is ok for me too.
And again, if you want, i can make a PR, I just need to have a providers folder, and a go `struct` to follow for implement features.

**Comment by ckotzbauer** [comment]
I'm open for PRs. But first we need to abstract the kubernetes-logic away and define the interface-structs before we can start implementing new orchestrators. It would be simpler if I would know your implementation for nomad. Can you share your gitlab repo privately?

**Comment by maxlerebourg** [comment]
I implemented also the cron /daemon too, but I 'm facing a weird behavior.
I have the git repository with
```
+ schollz
   + cowyo_latest
      + sbom.json
+ traefik
   + whoami
      + sbom.json
```
Normally it's pretty classic.
But when I log inside the `processor.executeSyftScans` the two parameters inside the `if containsImage(allImages, t.ImageID)`, with `logrus.Infof(t.ImageID, allImages)` I have 
```
traefik/whoami, [%!s(*oci.RegistryImage=&{traefik/whoami traefik/whoami []}) %!s(*oci.RegistryImage=&{schollz/cowyo:latest schollz/cowyo:latest []})]
schollz/cowyo_latest, [%!s(*oci.RegistryImage=&{traefik/whoami traefik/whoami []}) %!s(*oci.RegistryImage=&{schollz/cowyo:latest schollz/cowyo:latest []})]
Image schollz/cowyo_latest marked for removal
```
I don't know if it's about my change to the repo, but when you `target.LoadImages()` from git, the list of string is formatted as folder `schollz:latest => schollz_latest` and in conclusion all the image with a specific version is marked as deleted.

**Comment by ckotzbauer** [comment]
I'm currently working on a refactoring, making it possible to opt-out kubernetes. I started integrating your nomad code (as of  last Tuesday)
The image removal logic is pretty stable, I think. So maybe it's related to your changes, but I'm not sure.
